File: AlgorithmsProblemSolvingJava/src/io/programminglife/leetcode/Arrays/MajorityElement.java
package io.programminglife.leetcode.Arrays;

import java.util.HashMap;
import java.util.Map;

/**
 * Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.
 * You may assume that the array is non-empty and the majority element always exist in the array.
 */
public class MajorityElement {

    public static int majorityElement(int[] nums) {
        int majority = 0;
        Map<Integer, Integer> countMap = new HashMap<>();

        for(int i = 0; i < nums.length; i++) {
            if(countMap.get(nums[i]) != null) {
                int tempCount = countMap.get(nums[i]) + 1;
                countMap.put(nums[i], tempCount);
            } else {
                countMap.put(nums[i], 1);
            }
        }
        
        int tempMajority = 0;
        for(int key : countMap.keySet()) {
            if(countMap.get(key) > tempMajority) {
                tempMajority = countMap.get(key);
                majority = key;
            }
        }

        return majority;
    }

    public static void main(String[] args) {
        int[] nums = {2,2,1,1,1,2,2};

        System.out.println(majorityElement(nums));
    }
    
}

File: AlgorithmsProblemSolvingPython/leetcode/Arrays/majorityElement.py
def majority_element(nums):
    majority = 0
    countMap = {}

    for i in range(len(nums)):
        if countMap.get(nums[i]):
            countMap[nums[i]] += 1
        else: 
            countMap[nums[i]] = 1

    temp = 0
    for key in countMap.keys():
        if countMap[key] > temp:
            temp = countMap[key]
            majority = key

    return majority

print(majority_element([3,1,3]))